### PR TITLE
Parallel flush of pending INSERT blocks of Distributed engine

### PR DESCRIPTION
**Pull request by azat**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Parallel flush of pending INSERT blocks of Distributed engine on `DETACH`/server shutdown and `SYSTEM FLUSH DISTRIBUTED` (Parallelism will work only if you have multi disk policy for table (like everything in Distributed engine right now))

### CODE REVIEW

**Comment by azat** [comment]
Test failures does not looks related:

>Stateless tests (ubsan) [2/2] — Tests are not finished, fail: 9, passed: 668, skipped: 7

```
2024.02.23 17:10:40.127400 [ 1375 ] {} <Fatal> : Logical error: 'Part all_73_73_0 remains in ZooKeeper after DROP_RANGE all_0_73_999999999_999999999'.
2024.02.23 17:21:34.839947 [ 25932 ] {} <Fatal> BaseDaemon: ########## Short fault info ############
2024.02.23 17:21:34.847283 [ 25932 ] {} <Fatal> BaseDaemon: (version 24.2.1.1, build id: 9B05F21CAB88320B6D598A91C79A1BE778FD0582, git hash: fb55343088ec89c23ebe6b4f623cdc0c93148446) (from thread 1375) Received signal 6
2024.02.23 17:21:34.847323 [ 25932 ] {} <Fatal> BaseDaemon: Signal description: Aborted
2024.02.23 17:21:34.847343 [ 25932 ] {} <Fatal> BaseDaemon: 
2024.02.23 17:21:34.847370 [ 25932 ] {} <Fatal> BaseDaemon: Stack trace: 0x00007efdecf4f9fc 0x00007efdecefb476 0x00007efdecee17f3 0x00005561679bfec8 0x00005561679c0e15 0x000055615b67f9ea 0x00005561764add29 0x000055617647b1f0 0x00005561764738d6 0x00005561765cec5f 0x0000556176fd9e7b 0x00005561764d58f2 0x0000556176600279 0x0000556176b09c2f 0x0000556176b0a8a4 0x0000556167ac02bb 0x0000556167ac5314 0x0000556167abd91b 0x0000556167ac266a 0x00007efdecf4dac3 0x00007efdecfdea04
2024.02.23 17:21:34.847393 [ 25932 ] {} <Fatal> BaseDaemon: ########################################
2024.02.23 17:21:34.847417 [ 25932 ] {} <Fatal> BaseDaemon: (version 24.2.1.1, build id: 9B05F21CAB88320B6D598A91C79A1BE778FD0582, git hash: fb55343088ec89c23ebe6b4f623cdc0c93148446) (from thread 1375) (no query) Received signal Aborted (6)
2024.02.23 17:21:34.847436 [ 25932 ] {} <Fatal> BaseDaemon: 
2024.02.23 17:21:34.847454 [ 25932 ] {} <Fatal> BaseDaemon: Stack trace: 0x00007efdecf4f9fc 0x00007efdecefb476 0x00007efdecee17f3 0x00005561679bfec8 0x00005561679c0e15 0x000055615b67f9ea 0x00005561764add29 0x000055617647b1f0 0x00005561764738d6 0x00005561765cec5f 0x0000556176fd9e7b 0x00005561764d58f2 0x0000556176600279 0x0000556176b09c2f 0x0000556176b0a8a4 0x0000556167ac02bb 0x0000556167ac5314 0x0000556167abd91b 0x0000556167ac266a 0x00007efdecf4dac3 0x00007efdecfdea04
2024.02.23 17:21:34.847487 [ 25932 ] {} <Fatal> BaseDaemon: 3. ? @ 0x00007efdecf4f9fc
2024.02.23 17:21:34.847507 [ 25932 ] {} <Fatal> BaseDaemon: 4. ? @ 0x00007efdecefb476
2024.02.23 17:21:34.847564 [ 25932 ] {} <Fatal> BaseDaemon: 5. ? @ 0x00007efdecee17f3
2024.02.23 17:21:34.881187 [ 25932 ] {} <Fatal> BaseDaemon: 6. ./build_docker/./src/Common/Exception.cpp:43: DB::abortOnFailedAssertion(String const&) @ 0x0000000025f19ec8 in /usr/bin/clickhouse
2024.02.23 17:21:34.921779 [ 25932 ] {} <Fatal> BaseDaemon: 7.0. inlined from ./build_docker/./src/Common/Exception.cpp:63: DB::handle_error_code(String const&, int, bool, std::vector<void*, std::allocator<void*>> const&)
2024.02.23 17:21:34.921815 [ 25932 ] {} <Fatal> BaseDaemon: 7. ./build_docker/./src/Common/Exception.cpp:100: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x0000000025f1ae15 in /usr/bin/clickhouse
2024.02.23 17:25:34.316284 [ 25932 ] {} <Fatal> BaseDaemon: 8. DB::Exception::Exception<String const&, String const&>(int, FormatStringHelperImpl<std::type_identity<String const&>::type, std::type_identity<String const&>::type>, String const&, String const&) @ 0x0000000019bd99ea in /usr/bin/clickhouse
2024.02.23 17:25:34.636670 [ 25932 ] {} <Fatal> BaseDaemon: 9. ./build_docker/./src/Storages/StorageReplicatedMergeTree.cpp:2281: DB::paranoidCheckForCoveredPartsInZooKeeper(std::shared_ptr<zkutil::ZooKeeper> const&, String const&, StrongTypedef<unsigned int, DB::MergeTreeDataFormatVersionTag>, String const&) @ 0x0000000034a07d29 in /usr/bin/clickhouse
2024.02.23 17:25:34.840536 [ 25932 ] {} <Fatal> BaseDaemon: 10.0. inlined from ./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:701: ~shared_ptr
2024.02.23 17:25:34.840578 [ 25932 ] {} <Fatal> BaseDaemon: 10. ./build_docker/./src/Storages/StorageReplicatedMergeTree.cpp:2344: DB::StorageReplicatedMergeTree::executeDropRange(DB::ReplicatedMergeTreeLogEntry const&) @ 0x00000000349d51f0 in /usr/bin/clickhouse
2024.02.23 17:25:35.044287 [ 25932 ] {} <Fatal> BaseDaemon: 11. ./build_docker/./src/Storages/StorageReplicatedMergeTree.cpp:0: DB::StorageReplicatedMergeTree::executeLogEntry(DB::ReplicatedMergeTreeLogEntry&) @ 0x00000000349cd8d6 in /usr/bin/clickhouse
2024.02.23 17:25:35.356866 [ 25932 ] {} <Fatal> BaseDaemon: 12. ./contrib/llvm-project/libcxx/include/__functional/function.h:716: ? @ 0x0000000034b28c5f in /usr/bin/clickhouse
2024.02.23 17:25:35.459624 [ 25932 ] {} <Fatal> BaseDaemon: 13. ./contrib/llvm-project/libcxx/include/__functional/function.h:0: ? @ 0x0000000035533e7b in /usr/bin/clickhouse
2024.02.23 17:25:35.604823 [ 25932 ] {} <Fatal> BaseDaemon: 14.0. inlined from ./contrib/llvm-project/libcxx/include/__functional/function.h:818: ?
2024.02.23 17:25:35.604854 [ 25932 ] {} <Fatal> BaseDaemon: 14.1. inlined from ./contrib/llvm-project/libcxx/include/__functional/function.h:1174: ?
2024.02.23 17:25:35.604869 [ 25932 ] {} <Fatal> BaseDaemon: 14. ./build_docker/./src/Storages/StorageReplicatedMergeTree.cpp:3508: DB::StorageReplicatedMergeTree::processQueueEntry(std::shared_ptr<DB::ReplicatedMergeTreeQueue::SelectedEntry>) @ 0x0000000034a2f8f2 in /usr/bin/clickhouse
2024.02.23 17:25:35.913862 [ 25932 ] {} <Fatal> BaseDaemon: 15.0. inlined from ./contrib/llvm-project/libcxx/include/__functional/function.h:848: ?
2024.02.23 17:25:35.913885 [ 25932 ] {} <Fatal> BaseDaemon: 15.1. inlined from ./contrib/llvm-project/libcxx/include/__functional/function.h:1187: ?
2024.02.23 17:25:35.913898 [ 25932 ] {} <Fatal> BaseDaemon: 15. ./src/Storages/MergeTree/IExecutableTask.h:70: DB::ExecutableLambdaAdapter::executeStep() @ 0x0000000034b5a279 in /usr/bin/clickhouse
2024.02.23 17:25:35.925225 [ 25932 ] {} <Fatal> BaseDaemon: 16. ./build_docker/./src/Storages/MergeTree/MergeTreeBackgroundExecutor.cpp:294: DB::MergeTreeBackgroundExecutor<DB::RoundRobinRuntimeQueue>::routine(std::shared_ptr<DB::TaskRuntimeData>) @ 0x0000000035063c2f in /usr/bin/clickhouse
2024.02.23 17:25:35.936487 [ 25932 ] {} <Fatal> BaseDaemon: 17.0. inlined from ./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:701: ~shared_ptr
2024.02.23 17:25:35.936506 [ 25932 ] {} <Fatal> BaseDaemon: 17. ./build_docker/./src/Storages/MergeTree/MergeTreeBackgroundExecutor.cpp:349: DB::MergeTreeBackgroundExecutor<DB::RoundRobinRuntimeQueue>::threadFunction() @ 0x00000000350648a4 in /usr/bin/clickhouse
2024.02.23 17:25:35.947757 [ 25932 ] {} <Fatal> BaseDaemon: 18.0. inlined from ./base/base/../base/wide_integer_impl.h:810: bool wide::integer<128ul, unsigned int>::_impl::operator_eq<wide::integer<128ul, unsigned int>>(wide::integer<128ul, unsigned int> const&, wide::integer<128ul, unsigned int> const&)
2024.02.23 17:25:35.947779 [ 25932 ] {} <Fatal> BaseDaemon: 18.1. inlined from ./base/base/../base/wide_integer_impl.h:1483: bool wide::operator==<128ul, unsigned int, 128ul, unsigned int>(wide::integer<128ul, unsigned int> const&, wide::integer<128ul, unsigned int> const&)
2024.02.23 17:25:35.947790 [ 25932 ] {} <Fatal> BaseDaemon: 18.2. inlined from ./base/base/../base/strong_typedef.h:42: StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag>::operator==(StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag> const&) const
2024.02.23 17:25:35.947807 [ 25932 ] {} <Fatal> BaseDaemon: 18.3. inlined from ./src/Common/OpenTelemetryTraceContext.h:65: DB::OpenTelemetry::Span::isTraceEnabled() const
2024.02.23 17:25:35.947819 [ 25932 ] {} <Fatal> BaseDaemon: 18. ./build_docker/./src/Common/ThreadPool.cpp:457: ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::worker(std::__list_iterator<ThreadFromGlobalPoolImpl<false>, void*>) @ 0x000000002601a2bb in /usr/bin/clickhouse
2024.02.23 17:25:35.957310 [ 25932 ] {} <Fatal> BaseDaemon: 19. ./build_docker/./src/Common/ThreadPool.cpp:0: ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()::operator()() @ 0x000000002601f314 in /usr/bin/clickhouse
2024.02.23 17:25:35.967350 [ 25932 ] {} <Fatal> BaseDaemon: 20.0. inlined from ./base/base/../base/wide_integer_impl.h:810: bool wide::integer<128ul, unsigned int>::_impl::operator_eq<wide::integer<128ul, unsigned int>>(wide::integer<128ul, unsigned int> const&, wide::integer<128ul, unsigned int> const&)
2024.02.23 17:25:35.967376 [ 25932 ] {} <Fatal> BaseDaemon: 20.1. inlined from ./base/base/../base/wide_integer_impl.h:1483: bool wide::operator==<128ul, unsigned int, 128ul, unsigned int>(wide::integer<128ul, unsigned int> const&, wide::integer<128ul, unsigned int> const&)
2024.02.23 17:25:35.967387 [ 25932 ] {} <Fatal> BaseDaemon: 20.2. inlined from ./base/base/../base/strong_typedef.h:42: StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag>::operator==(StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag> const&) const
2024.02.23 17:25:35.967397 [ 25932 ] {} <Fatal> BaseDaemon: 20.3. inlined from ./src/Common/OpenTelemetryTraceContext.h:65: DB::OpenTelemetry::Span::isTraceEnabled() const
2024.02.23 17:25:35.967408 [ 25932 ] {} <Fatal> BaseDaemon: 20. ./build_docker/./src/Common/ThreadPool.cpp:457: ThreadPoolImpl<std::thread>::worker(std::__list_iterator<std::thread, void*>) @ 0x000000002601791b in /usr/bin/clickhouse
2024.02.23 17:25:35.980326 [ 25932 ] {} <Fatal> BaseDaemon: 21.0. inlined from ./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:302: std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>::reset[abi:v15000](std::__thread_struct*)
2024.02.23 17:25:35.980346 [ 25932 ] {} <Fatal> BaseDaemon: 21.1. inlined from ./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:259: ~unique_ptr
2024.02.23 17:25:35.980355 [ 25932 ] {} <Fatal> BaseDaemon: 21.2. inlined from ./contrib/llvm-project/libcxx/include/tuple:265: ~__tuple_leaf
2024.02.23 17:25:35.980364 [ 25932 ] {} <Fatal> BaseDaemon: 21.3. inlined from ./contrib/llvm-project/libcxx/include/tuple:538: ~tuple
2024.02.23 17:25:35.980381 [ 25932 ] {} <Fatal> BaseDaemon: 21.4. inlined from ./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:48: std::default_delete<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void ThreadPoolImpl<std::thread>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>>::operator()[abi:v15000](std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void ThreadPoolImpl<std::thread>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>*) const
2024.02.23 17:25:35.980398 [ 25932 ] {} <Fatal> BaseDaemon: 21.5. inlined from ./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:305: std::unique_ptr<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void ThreadPoolImpl<std::thread>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>, std::default_delete<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void ThreadPoolImpl<std::thread>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>>>::reset[abi:v15000](std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void ThreadPoolImpl<std::thread>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>*)
2024.02.23 17:25:35.980416 [ 25932 ] {} <Fatal> BaseDaemon: 21.6. inlined from ./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:259: ~unique_ptr
2024.02.23 17:25:35.980428 [ 25932 ] {} <Fatal> BaseDaemon: 21. ./contrib/llvm-project/libcxx/include/thread:297: void* std::__thread_proxy[abi:v15000]<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void ThreadPoolImpl<std::thread>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>>(void*) @ 0x000000002601c66a in /usr/bin/clickhouse
2024.02.23 17:25:35.980439 [ 25932 ] {} <Fatal> BaseDaemon: 22. ? @ 0x00007efdecf4dac3
2024.02.23 17:25:35.980447 [ 25932 ] {} <Fatal> BaseDaemon: 23. ? @ 0x00007efdecfdea04
2024.02.23 17:25:35.980458 [ 25932 ] {} <Fatal> BaseDaemon: Integrity check of the executable skipped because the reference checksum could not be read.
2024.02.23 17:25:38.574702 [ 25932 ] {} <Fatal> BaseDaemon: This ClickHouse version is not official and should be upgraded to the official build.
```

- https://github.com/ClickHouse/ClickHouse/issues/56037

>Integration tests (tsan) [3/6] — fail: 1, passed: 335

- `test_storage_azure_blob_storage/test.py::test_union_schema_inference_mode`

> AST fuzzer (asan) — Logical error: 'Expected single dictionary argument for function.'.

- https://github.com/ClickHouse/ClickHouse/issues/58979

> AST fuzzer (msan) — ERROR: MemorySanitizer: requested allocation size 0xc00000000 exceeds maximum supported size of 0x200000000

- https://github.com/ClickHouse/ClickHouse/issues/60352

File: src/Storages/StorageDistributed.cpp [hunk 1]
```diff
@@ -1265,10 +1265,19 @@ void StorageDistributed::initializeFromDisk()
     /// Make initialization for large number of disks parallel.
     ThreadPool pool(CurrentMetrics::StorageDistributedThreads, CurrentMetrics::StorageDistributedThreadsActive, CurrentMetrics::StorageDistributedThreadsScheduled, disks.size());
 
+    ThreadGroupPtr thread_group = CurrentThread::getGroup();
     for (const DiskPtr & disk : disks)
     {
         pool.scheduleOrThrowOnError([&]()
         {
+            SCOPE_EXIT_SAFE(
```
Comment (antonio2368) [inline src/Storages/StorageDistributed.cpp:1273]: you can probably use `scheduleFromThreadPool` from `threadPoolCallbackRunner.h`

File: src/Storages/StorageDistributed.cpp [hunk 2]
```diff
@@ -1735,16 +1752,36 @@ void StorageDistributed::flushClusterNodesAllData(ContextPtr local_context)
...
-    /// TODO: Maybe it should be executed in parallel
-    for (auto & node : directory_queues)
+        Stopwatch watch;
+        ThreadGroupPtr thread_group = CurrentThread::getGroup();
+        ThreadPool pool(CurrentMetrics::StorageDistributedThreads, CurrentMetrics::StorageDistributedThreadsActive, CurrentMetrics::StorageDistributedThreadsScheduled, directory_queues.size());
+        for (const auto & node : directory_queues)
+        {
+            pool.scheduleOrThrowOnError([node_to_flush = node, &thread_group]()
+            {
+                SCOPE_EXIT_SAFE(
```
Comment (antonio2368) [inline src/Storages/StorageDistributed.cpp:1766]: same as above, maybe use `scheduleFromThreadPool` and wait for futures here

File: src/Storages/StorageDistributed.cpp (hunk 1 above)
Comment (azat) [reply src/Storages/StorageDistributed.cpp]: Done

File: src/Storages/StorageDistributed.cpp (hunk 2 above)
Comment (azat) [reply src/Storages/StorageDistributed.cpp]: Make sense, thanks, done.

File: src/Storages/StorageDistributed.cpp [hunk 3]
```diff
@@ -1286,25 +1299,28 @@ void StorageDistributed::initializeFromDisk()
 
     /// Make initialization for large number of disks parallel.
     ThreadPool pool(CurrentMetrics::StorageDistributedThreads, CurrentMetrics::StorageDistributedThreadsActive, CurrentMetrics::StorageDistributedThreadsScheduled, disks.size());
+    std::vector<std::future<void>> futures;
 
     for (const DiskPtr & disk : disks)
     {
-        pool.scheduleOrThrowOnError([&]()
+        auto future = scheduleFromThreadPool<void>([this, disk_to_init = disk]
```
Comment (antonio2368) [inline src/Storages/StorageDistributed.cpp:1306]: where do you add this `future` to `futures`? :thinking: 
same for other 2 places

Comment (azat) [reply src/Storages/StorageDistributed.cpp:1305]: :facepalm: 
Fixed.

File: src/Storages/StorageDistributed.cpp [hunk 4]
```diff
@@ -1286,25 +1298,30 @@ void StorageDistributed::initializeFromDisk()
...
         {
-            initializeDirectoryQueuesForDisk(disk);
-        });
+            initializeDirectoryQueuesForDisk(disk_to_init);
+        }, pool, "DistInit");
+        futures.push_back(future);
```
Comment (antonio2368) [inline src/Storages/StorageDistributed.cpp:1309]: I think you need to move

**Review by antonio2368 — APPROVED** [review]

**Pull request merged**